### DISCUSSION TITLE
Customizaton?

### DISCUSSION THREAD

**Discussion by massiveradiation** [vlachoudis/bCNC#1897] [Q&A]
Hey folks,
I've been using bCNC almost daily for almost 5 years now and love the software. I've had little issue and what I find sets it apart from other senders is the: super easy Z autoprobing, ability to quickly edit gcode by clicking in toolpath, macros, etc. 
There are some specific functions that I would like to add based on my manufacturing use case and I wonder if anyone on here is actively working on bCNC and would like to earn some money customizing the software for my needs. 
I don't know the etiquette on github regarding requests for paid work, so I hope I'm not breaking any rules. If there is an official way to do this, I'm all ears. 
Cheers!
Tyler

**Comment by Harvie** [comment]
There is nothing wrong with such approach. Maybe it will attract new developers to bCNC project. Can you please provide detailed description of what features you want to get in bCNC?

**Reply by massiveradiation** [reply]
Hi Harvie,
Thanks for getting back to me so quickly. I'm particularly interested in creating multiple zones for autoprobing. For example, I want to put three or more separate boards on the same CNC table, input their XY locations, and have auto probe automatically scan each specifide rectangle as it's own height map. Then open one gcode file and have the separate height map zones apply. This would help greatly in maximizing CNC space for repetitive manufacturing. In addition, these separate probe areas should remain in memory and should be toggleable (is that a word?)  Once the zone coordinates to be height mapped are entered and active, one click, walk away, and they all get probed. Once the gcode file is opened, one can apply the separate height maps to the gcode in the same way it's done now, just a single click on the target button. 
I think the complicated bit would be keeping interpolation between probing points only to the individual zones. 
Let me know if this makes sense. 
Also, I want to be able to assign keyboard keys to macro icons and allow more macros. Maybe this is currently possible, but I can't figure out how to do it.
I have a few more smaller things I'd like to add or change, but the first one is the most pressing for me. 
Anyway, anyone who thinks this is in their wheelehouse and willing to give it a whirl, there's some dinero in it for you.

**Comment by massiveradiation** [comment]
Harvie,
Are you interested in tackling this? I have a budget for this development.
Cheers,
Tyler

**Reply by Harvie** [reply]
I am rather busy, but i am thinking about some of the features you've said. Can't you just autoprobe whole workspace? that way you wouldn't need to have separate autoprobe maps for each part of the workspace.

**Comment by massiveradiation** [comment]
Hi Tomas,
Thanks for getting back to me!The problem with butting up two pieces would
be at the joint between them, which may be different thicknesses. Between
the row of probe points on one board to the next row on the other board,
the data would be interpreted as a slope and throw off any engraving close
to the edge. Since I'm engraving really fine detailed stuff, even a
variation of 0.1mm makes a noticable difference.
Ideally I would set as many as 4 different probe zones, which would each
have their own height map which could be applied to a single gcode file in
bCNC.
Are you the brain child behind bCNC? This software is so good!
If you don't have the bandwidth for your feature, perhaps there is one
small one you could take, which is assigning macro buttons to keys on the
keyboard. Or if this is already possible, I don't know how to do it.
Let me know if you can squeeze the height probe project into your schedule.
Cheers,
Tyler

**Comment by massiveradiation** [comment]
Sorry, I should never send an email before my first cup of coffee. I'll
write another one later that spells things out a little more clearly.
Cheers,
Tyler

**Comment by massiveradiation** [comment]
Hi Tomas,
Not sure if my reply went through. I think I should have hit "reply all".

Message below:

Hi Tomas,
Thanks for getting back to me! The problem with butting up two pieces would
be at the joint between them, which may be different thicknesses. Between
the row of probe points on one board and the other board, the data would be
interpolatwd as a slope and throw off any engraving close to the edge.
Since I'm engraving really fine detailed stuff, even a variation of 0.1mm
makes a noticable difference.
Ideally I would set as many as 4 different auto level "zones," which would
each have their own height map which could be applied to a single gcode
file in bCNC based on XY coordinates.
Are you the brain child behind bCNC? This software is so good!
If you don't have the bandwidth for this autolevel multi zone feature,
perhaps there is one small feature you would be able to add, which is
assigning macro buttons to keys on the keyboard. Or if this is already
possible, I don't know how to do it.
Let me know if you can squeeze the autolevel multizone project into your
schedule and we can talk details and payment
Cheers,
Tyler

**Comment by Mechscientist** [comment]
Hi Tyler,

I don't know if you got this resolved. I happened to stumble across this in a search. I'm a retired software developer and it sounded like an interesting puzzle (and solving it might increase my understanding of how bCNC works ;-). Thought I'd share my thoughts in case it's helpful to someone.

My understanding of the way auto leveling works is that there is a Probe 'object'  (in CNC.py). The Probe object manages the z-offset data and has the functionality for creating the g-code to probe the surface, interpolation of the position g-codes while running and save/store of the offsets in a ".probe" file. 

The leveling offsets are gathered by generating the g-code to move through a grid of XY points and sending a G38.X word to the machine controller. Apparently the machine controller responds from a probe command with the XYZ position at the point of contact. When the Sender receives this response, it tells the Probe object to save that location.

Armchair coding: 

I would approach this by creating an object that would manage multiple Probe objects, having the same 'interface' as (i.e. duck-typing) the Probe object. I'm going to call it a ProbeContainer, for lack of a better name. The ProbeContainer could then contain multiple Probe objects, each for an area of the work plane. For the Probe functions like 'add' and 'interpolate', the ProbeContainer object would compare the x and y of the input to each Probe's margins to determine which one to use and pass the arguments to it. 

Similarly, file save and load could just be done by looping over the Probe objects and calling save or load with some naming protocol. 

ProbeContainer would need to replace the Probe object created in the initializer of GCode. This presents a bit of a problem, but it could be because I'm just armchair reverse engineering. You could just go in to CNC.py and change the line of code in the GCode initializer that the Probe object is created to create a ProbeContainer instead (creating it with one Probe object), but if you upgrade bCNC, you will have to remember make that modification again. There may be some other places that you could substitute in a probe manager, like from a plugin that wouldn't be as intrusive, but perhaps someone more familiar with the code can comment on that.

The last part of the code required would be to create a user interface to input the locations of the probe areas. I have found that GUI's are a matter of preference,, but the easiest would be to just use the plugin framework. The ProbeContainer object would be accessible from the app as app.gcode.probe.

Cheers!

**Comment by Git-Bruno** [comment]
Maybe this is out of scope here but it would be nice to have something like a manual mode for auto level to do auto leveling on non conductive surfaces.
The idea would be to travel through all the probe points but pause at each probe points waiting for the operator to move the probe contact under the probe tool. Probing would continue after the operator confirms the probe contact is correctly positioned.

**Reply by Harvie** [reply]
Just mount some kind of mechanical switch to the chuck:

https://www.youtube.com/watch?v=dgqRBbZXwIA&t=44s
https://www.thingiverse.com/thing:721620

**Reply by Mechscientist** [reply]
I recently purchased one of these:
[https://www.amazon.com/PGFUN-Anti-roll-Waterproof-Coordinate-Compatible/dp/B0B3XMFDX5/ref=sr_1_2_sspa?dib=eyJ2IjoiMSJ9.mETHK55cQ6rm_yd3kZRo98VIGSMXjzAuO5bUu_VY8__KopOkaAEGSLQivK9EMRwgyhdgXbo8jHaqPZLLcOIhEESSWZyqrai7yN9wXW763LRaQZyoEbMa_0U2mqJxSmGbDT19rJIZiX-I-7by4hLnMa-jhYJYUDH3mHQsgwDkZELfBn3_ovVziNmlpOUg8u3t7zp4VGq9Vl9DnWrVLKfyHXhBuaIuxbniZBuSw_YlCVU.kejy0cb2DQyL7ZYn2FmkrbajXqL4wvPxDIOEcYxQ5oQ&dib_tag=se&keywords=3d%2Btouch%2Bprobe&qid=1737562631&sr=8-2-spons&sp_csd=d2lkZ2V0TmFtZT1zcF9hdGY&th=1](url)
For the price, it's quite repeatable, and doesn't require a conductive surface. It also works for setting up the X/Y coordinates as well.

**Reply by Git-Bruno** [reply]
This is exactly what I need.
Thanks to you 2.

**Comment by massiveradiation** [comment]
Hello,
Sounds like you'd be intrigued enough to develop some bCNC features for me?
I'm not a programmer, but I do understand the basic concepts and write
macros for bCNC. Can you contact me directly to discuss?
Cheers,
Tyler
